### PR TITLE
P2167R3 Improved Proposed Wording for LWG 2114 (contextually converti…

### LINKED ISSUE #5971: [2022-11 LWG Motion 10] P2167R3 Improved Proposed Wording for LWG 2114 (contextually convertible to bool)
https://wg21.link/p2167r3

### PR DESCRIPTION
**Pull request by jensmaurer**
…ble to bool)

Fixes NB US 073 (C++23 CD).

Fixes #5971
Fixes cplusplus/papers#877
Fixes cplusplus/nbballot#504

### CODE REVIEW

File: source/threads.tex [hunk 1]
```diff
@@ -30,8 +30,14 @@
 \pnum
 Throughout this Clause, the names of template parameters are used to express type
 requirements.
-If a template parameter is named \tcode{Predicate}, \tcode{operator()} applied to
-the template argument shall return a value that is convertible to \tcode{bool}.
+\tcode{Predicate} is a function object type\iref{function.objects}.
+Let \tcode{pred} denote an lvalue of type \tcode{Predicate}.
+Then the expression \tcode{pred()} shall be well-formed and
+the type \tcode{decltype(pred())} shall model
+\exposconcept{boolean-testable}\iref{concept.booleantestable]}.
```
Comment (JohelEGP) [inline source/threads.tex:37]:
```suggestion
\exposconcept{boolean-testable}\iref{concept.booleantestable}.
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/threads.tex (hunk 1 above)
Comment (jensmaurer) [reply source/threads.tex]: Fixed

**Review by burblebee — CHANGES_REQUESTED** [review]
2 issues: (1) widen "Return type" column in tables and (2) fix wording for "each [...] models".

File: source/support.tex [hunk 2]
```diff
@@ -5045,7 +5045,9 @@
   Otherwise, \tcode{strong_order(E, F)} if it is a well-formed expression.
 \item
   Otherwise, if the expressions \tcode{E == F} and \tcode{E < F}
-  are both well-formed and convertible to \tcode{bool},
+  are both well-formed and
+  each of \tcode{decltype(E == F)} and \tcode{decltype(E < F)} models
+  \exposconcept{boolean-testable},
```
[the remark this answers is not in the document]
Comment (Dani-Hub) [reply source/support.tex:5050]: Can we please keep the "each" regardless the position? My interpretation of a previous review of @W-E-Brown was that the "each" makes the interpretation unambiguous. But maybe Walter would like to correct my possibly incorrect interpretation of his actually words used?

[the remark this answers is not in the document]
Comment (JohelEGP) [reply source/support.tex:5050]: > and change "models" to "model" ("each" can be either plural or singular in this case, but plural sounds better and I think is what is most often used elsewhere in the spec).

According to https://github.com/cplusplus/draft/pull/5986#discussion_r1025760788, when "each" is next to "model", there are 10 occurrences for "model" and 0 for "models".

**Comment by W-E-Brown** [comment]
> On Nov 19, 2022, at 6:21 AM, Daniel Krügler ***@***.***> wrote:
> @Dani-Hub commented on this pull request.
> 
> In source/support.tex <https://github.com/cplusplus/draft/pull/5986#discussion_r1027083394>:
> 
> > @@ -5045,7 +5045,9 @@
>    Otherwise, \tcode{strong_order(E, F)} if it is a well-formed expression.
>  \item
>    Otherwise, if the expressions \tcode{E == F} and \tcode{E < F}
> -  are both well-formed and convertible to \tcode{bool},
> +  are both well-formed and
> +  each of \tcode{decltype(E == F)} and \tcode{decltype(E < F)} models
> +  \exposconcept{boolean-testable},
> Can we please keep the "each" regardless the position? My interpretation of a previous review of @W-E-Brown <https://github.com/W-E-Brown> was that the "each" makes the interpretation unambiguous. But maybe Walter would like to correct my possibly incorrect interpretation of his actually words used?
> 
Daniel has the right idea:  The words "each" and "both" are not interchangeable; they mean different things.  The use of "both" in this sentence changes the meaning, leading to an incorrect specification, as was discussed before the wording was approved in committee.

**Comment by W-E-Brown** [comment]
> On Nov 19, 2022, at 7:24 AM, Johel Ernesto Guerrero Peña ***@***.***> wrote:
> @JohelEGP commented on this pull request.
> 
> In source/support.tex <https://github.com/cplusplus/draft/pull/5986#discussion_r1027089740>:
> 
> > @@ -5045,7 +5045,9 @@
>    Otherwise, \tcode{strong_order(E, F)} if it is a well-formed expression.
>  \item
>    Otherwise, if the expressions \tcode{E == F} and \tcode{E < F}
> -  are both well-formed and convertible to \tcode{bool},
> +  are both well-formed and
> +  each of \tcode{decltype(E == F)} and \tcode{decltype(E < F)} models
> +  \exposconcept{boolean-testable},
> and change "models" to "model" ("each" can be either plural or singular in this case, but plural sounds better and I think is what is most often used elsewhere in the spec).
> 
> According to #5986 (comment) <https://github.com/cplusplus/draft/pull/5986#discussion_r1025760788>, when "each" is next to "model", there are 10 occurrences for "model" and 0 for "models".
> 
Sorry, but for such purposes "next to" appears to be an irrelevant grammatical criterion:  The sole applicable criterion, according to multiple dictionaries and grammar books, is whether the subject of the clause is singular or plural.
In Daniel's "each of ..." wording formulation, above, "each" is a pronoun used as the subject and is clearly singular, so "models" would be correct.
In the formulation "the expressions ... and ... each ...", the use of "each" is adverbial (meaning "individually"), rather than pronominal; this leaves "the expressions" as the plural subject, so "model" would be correct.
(Those above-cited "10 occurrences" should be probably be similarly audited for correctness via analogous analysis.)

**Comment by JohelEGP** [comment]
> In the formulation "the expressions ... and ... each ...", the use of "each" is adverbial (meaning "individually"), rather than pronominal; this leaves "the expressions" as the plural subject, so "model" would be correct. (Those above-cited "10 occurrences" should be probably be similarly audited for correctness via analogous analysis.)

The 10 occurrences are of the form "`T1` and `T2` each model `C`", so they should be correct.

**Comment by JohelEGP** [comment]
Here's the actual output:
```cli
$ grep 'each model ' *.tex
algorithms.tex:   \tcode{R1} and \tcode{R2} each model \libconcept{sized_range} and
algorithms.tex:\tcode{R1} and \tcode{R2} each model \libconcept{sized_range}, and
concepts.tex:that each model \exposconcept{boolean-testable-impl},
ranges.tex:  each model \libconceptx{for\-ward_range}{forward_range}, then \tcode{iterator_concept} denotes
ranges.tex:each model \exposconcept{bidirectional-common},
ranges.tex:\exposid{Base} and \exposid{InnerBase} each model \libconcept{forward_range},
ranges.tex:\exposid{Base} and \exposid{InnerBase} each model \libconcept{forward_range}.
ranges.tex:each model \tcode{\libconcept{derived_from}<bidirectional_iterator_category>}
ranges.tex:each model \libconcept{common_range},
ranges.tex:each model \tcode{\libconcept{derived_from}<forward_iterator_tag>},
```

File: source/utilities.tex [hunk 3]
```diff
@@ -2903,11 +2908,15 @@
...
 where $0 \leq \tcode{i} < \tcode{sizeof...(TTypes)}$,
-\tcode{get<i>(t) == get<i>(u)} is a valid expression
-returning a type that is convertible to \tcode{bool}.
+\tcode{get<i>(t) == get<i>(u)} is a valid expression.
 \tcode{sizeof...(TTypes)} equals
 \tcode{tuple_size_v<UTuple>}.
 
+\pnum
+\expects
+For all \tcode{i}, each \tcode{decltype(get<i>(t) == get<i>(u))} models
```
[the remark this answers is not in the document]
Comment (JohelEGP) [reply source/utilities.tex:2917]: To clarify, all wording around `tuple` of the form "for each _index_,  _use-of-index_ _description-of-use-of-index_" seem to omit "each".
https://github.com/cplusplus/draft/blob/fb0df97e7d7c2c75bbdf7164c33b0796024ff6d3/source/utilities.tex#L2145-L2146
Likewise for all `tuple`'s `operator=`.
> [2](https://eel.is/c++draft/tuple.rel#2)
[#](http://github.com/Eelis/draft/tree/9a4a02f8a10f75675094dc0f81e089d1751ff3d8/source/utilities.tex#L2900)
Mandates: For all i, where 0≤i<sizeof...(TTypes), get<i>(t) == get<i>(u) is a valid expression returning a type that is convertible to bool[.](https://eel.is/c++draft/tuple.rel#2.sentence-1)
sizeof...(TTypes) equals tuple_­size_­v<UTuple>[.](https://eel.is/c++draft/tuple.rel#2.sentence-2)
> -- https://eel.is/c++draft/tuple.rel#2

**Comment by burblebee** [comment]
When using "Each of", "each" is singular (so "models" is correct).
When using "x and y each", "each" is plural (so "model" is correct)..
Are there other cases that don't fit either of these patterns?

See https://www.masterclass.com/articles/is-each-singular-or-plural and similar sites.

> (Those above-cited "10 occurrences" should be probably be similarly audited for correctness via analogous analysis.)

They are correct as is.

File: source/support.tex (hunk 2 above)
[the remark this answers is not in the document]
Comment (jwakely) [reply source/support.tex:5050]: IMHO "decltype(E == F) and decltype(E < F) model boolean-testable" has a very different meaning from "decltype(E == F) and decltype(E < F) each model boolean-testable".

The former means `boolean-testable<decltype(E==F), decltype(E<F)>` and the latter means `boolean_testable<decltype(E==F)> && boolean-testable<decltype(E<F)>` which is very different. The latter is what we want.

**Review by jwakely — APPROVED** [review]
It might be nice if we consistently say **either** "A and B each model C" **or** "each of A and B models C", but the changes as applied are not incorrect. And as I noted in a comment, changing to "A and B model C" would definitely be incorrect, so maybe best not to fiddle with it!

File: source/utilities.tex (hunk 3 above)
[the remark this answers is not in the document]
Comment (tkoeppe) [reply source/utilities.tex:2921]: Thanks, sounds reasonable!

[the remark this answers is not in the document]
Comment (JohelEGP) [reply source/utilities.tex:2921]:
```suggestion
For all \tcode{i}, \tcode{decltype(get<i>(t) == get<i>(u))} models
```

File: source/support.tex (hunk 2 above)
[the remark this answers is not in the document]
Comment (tkoeppe) [reply source/support.tex:5052]: Thank you for the discussion! I won't make any changes here, please file separate issues if you think the wording should be improved. This looks subtle, and we should only deviate from previous LWG review with a proper discussion.

File: source/utilities.tex (hunk 3 above)
[the remark this answers is not in the document]
Comment (JohelEGP) [reply source/utilities.tex:2921]:
```suggestion
For each \tcode{i}, \tcode{decltype(get<i>(t) == get<i>(u))} models
```

[the remark this answers is not in the document]
Comment (tkoeppe) [reply source/utilities.tex]: I'm OK with "for all", that's a fairly common mathematical phrase.

File: source/lib-intro.tex [hunk 4]
```diff
@@ -1627,7 +1627,7 @@
 \topline
 \hdstyle{Expression}  &   \hdstyle{Return type} &   \rhdr{Requirement} \\ \capsep
 \tcode{a == b}  &
-convertible to \tcode{bool} &
+\tcode{decltype(a == b)} models \exposconceptx{boolean-testa\-ble}{boolean-testable} &
```
[the remark this answers is not in the document]
Comment (tkoeppe) [reply source/lib-intro.tex:1625]: We will replace these tables with @jensmaurer paragraph pattern, so this will not be a problem eventually.

**Pull request merged**